### ISSUE TITLE
npm module doesn't install new grunt plugin according to package.json update

### ISSUE THREAD

**Issue by visay** [ansible/ansible-modules-extras#957]
- **Issue Type**: Bug Report
- **Ansible Version**: ansible 1.9.2
- **Environment**: Ubuntu Server 14.04 LTS AMD64
- **Summary**: npm module doesn't install new grunt plugin according to package.json update.

**Steps To Reproduce**:
- Create a `package.json` and store it in a directory:

```
{
    "name": "Dummy",
    "version": "0.0.1",
    "devDependencies": {
        "grunt": "latest",
        "grunt-contrib-clean": "~0.6.0",
        "grunt-contrib-copy": "latest",
        "grunt-contrib-watch": "latest",
        "grunt-contrib-uglify": "latest",
        "grunt-contrib-compass": "latest",
        "grunt-contrib-cssmin": "~0.10.0",
        "grunt-contrib-concat": "~0.5.0"
    }
}
```
- Run ansible command to install them

```
ansible localhost -m npm -a "path=/path/to/my/dir/containing/package-json state=latest"
```

First run, everything is fine and I get `node_modules` directory with all plugins defined in `package.json`.
- Update `package.json` file adding `"grunt-contrib-imagemin": "~0.4.0"` and change version to `0.0.2`
- Call ansible command again

```
ansible localhost -m npm -a "path=/path/to/my/dir/containing/package-json state=latest"
```

**Result**: Nothing updated. imagemin plugin doesn't get installed
**Output**: Always get

```
localhost | success >> {
    "changed": false
}
```
- If runing `npm install` shell command manually inside the directory, it works fine and the new plugin for imagemin get downloaded.

**Comment by juhau1** [comment]
I'm having the same problem. This seems like a critical bug affecting many npm packages.

My main package.json says:

``` javascript
  "dependencies": {
    "my-sub-module": "latest"
  }
```

When updating my-sub-module package ansible npm module is not picking up the change, but continues with the old my-sub-module version.

Current workaround is to use shell module to do npm install as suggested by visay.

Ansible version: 1.9.3-1ppa~trusty

**Comment by robynbergeron** [comment]
[module: packaging/language/npm.py]

**Comment by robynbergeron** [comment]
@visay -- thanks for submitting this bug report, and thanks to @juhau1 as well for providing info around your ability to duplicate.

@chrishoffman -- pinging you to let you know about this bug report for the npm module.

**Comment by frinux** [comment]
Is this bug gonna be resolved? It is quite annoying (we have currently to remove the node_modules folder before npm install to be sure the last version is used...)
Thanks

**Comment by moises-silva** [comment]
@visay @juhau1 @frinux Probably the most important question here is what version of npm are you using? (npm --version)

TL;DR Please update using 'npm install npm -g' and do not rely on the version provided by your OS.

Regardless of the OS (Ubuntu, CentOS) what matters the most in this case is the npm version. The npm user interface seems to break all the time (changes to the output and/or in behaviour). Old npm versions seem to ignore "devDependencies" when doing 'npm update' and the output of 'npm outdated' also seems to have changed across npm versions (latest seem to support --json for easy parsing now). This ansible module relies on npm install, update, list and outdated a lot, so changes there can break this module with relative ease.

Really old npm versions seem FUBAR (1.x versions) so I'd advise to not even attempt use those with this ansible module, the only way to get them to work would be to execute "npm update" always regardless of anything else or doing a lot of work to check every package version individually rather than trusting 'npm outdated'.

I've faced similar issues to the ones described in this issue, but most of them disappeared when I did 'npm install npm -g' to update npm itself. One remaining issue was that even when all packages were up to date the ansible module would insist in doing 'npm update' and report changed state. I've fixed that and a few other issues in a new PR.

My recommendation is to first try to update npm, then try my branch if you still have issues and report back your feedback.

**Comment by visay** [comment]
Confirmed that this bug is fixed with npm update on Ubuntu 14.04.2 LTS (without checking out @moises-silva 's branch - I have ansible version 1.9.4).

`npm --version`

Output: `1.3.10`

`npm install npm -g`

`npm --version`

Output: `3.7.3`

Then follow the description in **Steps To Reproduce** above will get the expected result.

**Comment by springuper** [comment]
same problem with
- ansible 2.1.0.0
- Ubuntu 14.04.3 LTS
- npm 3.10.3

I add `webpack-md5-hash` to `devDependencies` and npm module report `changed: false`

**Comment by yshnb** [comment]
I'm facing same problem too. versions are below

- ansible 2.2.0.0
- Ubuntu 16.04.1 LTS
- npm 3.10.8

**Comment by jmvbxx** [comment]
I had the same problem and solved the following way. I have `npm` install via `apt` as always but then in a separate task I use the `npm` module to install `npm` before any other package and then it works properly.

It's not the nicest workaround but at least it keeps the work flowing.

- ubuntu 16.04
- ansible 2.3.0.0
- npm 4.5.0
